lib/puppet/context/trusted_information.rb

```diff
@@ -51,17 +61,42 @@ def self.remote(authenticated, node_name, certificate)
           [ext['oid'].freeze, ext['value'].freeze]
         end]
       end
-      new('remote', node_name, extensions)
+      new('remote', node_name, extensions, external)
     else
-      new(false, nil, {})
+      new(false, nil, {}, external)
     end
   end
 
   def self.local(node)
     # Always trust local data by picking up the available parameters.
     client_cert = node ? node.parameters['clientcert'] : nil
 
-    new('local', client_cert, {})
+    new('local', client_cert, {}, retrieve_trusted_external(client_cert))
+  end
+
+  def self.retrieve_trusted_external(certname)
+    deep_freeze(Puppet::TrustedExternal.indirection.find(certname))
+  end
+
+  # Deeply freezes the given object. The object and its content must be of the types:
+  # Array, Hash, Numeric, Boolean, Regexp, NilClass, or String. All other types raises an Error.
+  # (i.e. if they are assignable to Puppet::Pops::Types::Data type).
+  def self.deep_freeze(object)
+    case object
+    when Array
+      object.each {|v| deep_freeze(v) }
+      object.freeze
+    when Hash
+      object.each {|k, v| deep_freeze(k); deep_freeze(v) }
+      object.freeze
+    when NilClass, Numeric, TrueClass, FalseClass
+      # do nothing
+    when String
+      object.freeze
+    else
+      raise Puppet::Error, _("Unsupported data type: '%{klass}'") % { klass: object.class }
+    end
+    object
```

Comment:
I see this is copied from the parser scope. But puppet vendors `deep_merge`! I'm ok adding this duplication for now. I filed https://tickets.puppetlabs.com/browse/PUP-10014.

Reply:
I concede the point on code duplication - and I'd be up for pulling out the function to `Util` if this goes forward. Re Pup-10014, this is `deep_freeze`, not `deep_merge`.

Reply:
wow, total :brain fart: on my part